mcp/streamable.go

```diff
@@ -654,120 +701,370 @@ func (s *streamableClientConn) Read(ctx context.Context) (JSONRPCMessage, error)
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	case <-s.done:
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if s.err != nil {
+			return nil, s.err // Return explicit error if connection closed due to error
```

Comment:
This looks like a logically distinct change. Given the size and complexity of this PR, it would be good to separate out those changes that are unrelated to replay support into one or more separate CLs, with tests.